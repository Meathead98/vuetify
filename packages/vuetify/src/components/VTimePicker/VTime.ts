// Utils
import pad from '../VDatePicker/util/pad'

// Types
import Vue, { VNode } from 'vue'
import { PropValidator } from 'vue/types/options'

export const enum SelectMode {
  Hour = 'hour',
  Minute = 'minute',
  Second = 'second'
}

export type Format = 'ampm' | '24hr'
export type Period = 'am' | 'pm'
export type AllowFunction = (val: number) => boolean

interface Allowed {
  hours: AllowFunction | number[]
  minutes: AllowFunction | number[]
  seconds: AllowFunction | number[]
}

export interface Time {
  hour: number | null
  minute: number | null
  second: number | null
}

export function convert24to12 (hour: number) {
  return hour ? ((hour - 1) % 12 + 1) : 12
}

export function convert12to24 (hour: number, period: Period) {
  return hour % 12 + (period === 'pm' ? 12 : 0)
}

export function parseTime (value: string | null | Date): [Time, Period] {
  let hour, minute, second, period

  if (value == null || value === '') {
    hour = null
    minute = null
    second = null
    period = 'am'
  } else if (value instanceof Date) {
    hour = value.getHours()
    minute = value.getMinutes()
    second = value.getSeconds()
    period = hour > 12 ? 'pm' : 'am'
  } else {
    const [, h, m, , s, p] = value.trim().toLowerCase().match(/^(\d+):(\d+)(:(\d+))?([ap]m)?$/) || new Array(6)

    hour = p ? convert12to24(parseInt(h, 10), p as Period) : parseInt(h, 10)
    minute = parseInt(m, 10)
    second = parseInt(s || 0, 10)
    period = p || (hour >= 0 && hour < 12 ? 'am' : 'pm')
  }

  return [{ hour, minute, second }, period as Period]
}

export interface VTimeScopedProps {
  allowed: {
    hour: Function
    minute: Function
    second: Function
  }
  format: Format
  time: Time
  period: Period
  selectMode: SelectMode
  setPeriod: Function
  setTime: Function
  setSelectMode: Function
  useSeconds: boolean
}

export default Vue.extend({
  name: 'v-time',

  props: {
    allowed: {
      type: Object,
      default: () => ({ hours: () => true, minutes: () => true, seconds: () => true })
    } as PropValidator<Allowed>,
    format: {
      type: String,
      default: 'ampm',
      validator (val) {
        return ['ampm', '24hr'].includes(val)
      }
    } as PropValidator<Format>,
    min: String,
    max: String,
    value: {
      type: [String, Date],
      default: () => null
    },
    useSeconds: Boolean
  },

  data () {
    const [internalTime, period] = parseTime(this.value)

    return {
      period,
      selectMode: SelectMode.Hour,
      internalTime
    }
  },

  computed: {
    isAllowedHourCb (): AllowFunction {
      let cb: AllowFunction

      if (!this.allowed.hours) {
        cb = () => true
      } else if (this.allowed.hours instanceof Array) {
        cb = (val: number) => (this.allowed.hours as number[]).includes(val)
      } else {
        cb = this.allowed.hours
      }

      if (!this.min && !this.max) return cb

      const minHour = this.min ? Number(this.min.split(':')[0]) : 0
      const maxHour = this.max ? Number(this.max.split(':')[0]) : 23

      return (val: number) => {
        return val >= minHour * 1 &&
          val <= maxHour * 1 &&
          (!cb || cb(val))
      }
    },
    isAllowedMinuteCb (): AllowFunction {
      let cb: AllowFunction

      const isHourAllowed = !this.isAllowedHourCb || this.internalTime.hour === null || this.isAllowedHourCb(this.internalTime.hour)
      if (!this.allowed.minutes) {
        cb = () => true
      } else if (this.allowed.minutes instanceof Array) {
        cb = (val: number) => (this.allowed.minutes as number[]).includes(val)
      } else {
        cb = this.allowed.minutes
      }

      if (!this.min && !this.max) {
        return isHourAllowed ? cb : () => false
      }

      const [minHour, minMinute] = this.min ? this.min.split(':').map(Number) : [0, 0]
      const [maxHour, maxMinute] = this.max ? this.max.split(':').map(Number) : [23, 59]
      const minTime = minHour * 60 + minMinute * 1
      const maxTime = maxHour * 60 + maxMinute * 1

      return (val: number) => {
        const time = 60 * this.internalTime.hour! + val
        return time >= minTime &&
          time <= maxTime &&
          isHourAllowed &&
          (!cb || cb(val))
      }
    },
    isAllowedSecondCb (): AllowFunction {
      let cb: AllowFunction

      const isHourAllowed = !this.isAllowedHourCb || this.internalTime.hour === null || this.isAllowedHourCb(this.internalTime.hour)
      const isMinuteAllowed = isHourAllowed &&
        (!this.isAllowedMinuteCb ||
          this.internalTime.minute === null ||
          this.isAllowedMinuteCb(this.internalTime.minute)
        )

      if (!this.allowed.seconds) {
        cb = () => true
      } else if (this.allowed.seconds instanceof Array) {
        cb = (val: number) => (this.allowed.seconds as number[]).includes(val)
      } else {
        cb = this.allowed.seconds
      }

      if (!this.min && !this.max) {
        return isMinuteAllowed ? cb : () => false
      }

      const [minHour, minMinute, minSecond] = this.min ? this.min.split(':').map(Number) : [0, 0, 0]
      const [maxHour, maxMinute, maxSecond] = this.max ? this.max.split(':').map(Number) : [23, 59, 59]
      const minTime = minHour * 3600 + minMinute * 60 + (minSecond || 0) * 1
      const maxTime = maxHour * 3600 + maxMinute * 60 + (maxSecond || 0) * 1

      return (val: number) => {
        const time = 3600 * this.internalTime.hour! + 60 * this.internalTime.minute! + val
        return time >= minTime &&
          time <= maxTime &&
          isMinuteAllowed &&
          (!cb || cb(val))
      }
    },
    scopedSlotProps (): VTimeScopedProps {
      return {
        allowed: {
          hour: this.isAllowedHourCb,
          minute: this.isAllowedMinuteCb,
          second: this.isAllowedSecondCb
        },
        format: this.format,
        time: this.internalTime,
        period: this.period,
        selectMode: this.selectMode,
        setPeriod: this.setPeriod,
        setTime: this.setTime,
        setSelectMode: this.setSelectMode,
        useSeconds: this.useSeconds
      }
    },
    timeAsString (): string | null {
      const { hour, minute, second } = this.internalTime
      if (hour != null && minute != null && (!this.useSeconds || second != null)) {
        return `${pad(hour)}:${pad(minute)}` + (this.useSeconds ? `:${pad(second!)}` : '')
      }

      return null
    }
  },

  watch: {
    value (value: string | null | Date) {
      const [time, period] = parseTime(value)
      this.internalTime = time
      this.period = period
    },
    timeAsString (v: string | null) {
      if (v != null) this.$emit('input', v)
    },
    period (v: Period) {
      this.$emit('update:period', v)
    },
    selectMode (v: SelectMode) {
      this.$emit('update:selectMode', v)
    }
  },

  methods: {
    setPeriod (p: Period) {
      this.period = p

      // If hour is set, and we change period
      // then we need to update hour since it
      // is always stored in a 24 hour format
      if (this.internalTime.hour) {
        this.internalTime.hour = this.period === 'am' ? this.internalTime.hour - 12 : this.internalTime.hour + 12
      }
    },
    setTime (t: Time) { this.internalTime = t },
    setSelectMode (m: SelectMode) { this.selectMode = m }
  },

  render (): VNode {
    return this.$scopedSlots.default!(this.scopedSlotProps) as any
  }
})
