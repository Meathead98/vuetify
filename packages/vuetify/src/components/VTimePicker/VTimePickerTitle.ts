import './VTimePickerTitle.sass'

// Utils
import { pad } from '../VDatePicker/util'
import { genPickerButton } from '../VPicker/VPicker'

// Types
import Vue, { VNode } from 'vue'
import { PropValidator } from 'vue/types/options'
import { Time, SelectMode, Format, convert24to12 } from './VTime'

export default Vue.extend({
  name: 'v-time-picker-title',

  inheritAttrs: false,

  props: {
    format: {
      type: String,
      default: 'ampm'
    } as PropValidator<Format>,
    showAmPm: Boolean,
    disabled: Boolean,
    period: {
      type: String,
      validator: period => period === 'am' || period === 'pm'
    } as PropValidator<'am' | 'pm'>,
    readonly: Boolean,
    useSeconds: Boolean,
    selectMode: String as PropValidator<SelectMode>,
    time: Object as PropValidator<Time>
  },

  computed: {
    isAmPm (): boolean {
      return this.format === 'ampm'
    }
  },

  methods: {
    genTime () {
      let hour = this.time ? this.time.hour : null
      if (hour != null && this.isAmPm) {
        hour = convert24to12(hour)
      }

      const displayedHour = hour == null ? '--' : this.isAmPm ? String(hour) : pad(hour)
      const displayedMinute = this.time && this.time.minute != null ? pad(this.time.minute) : '--'
      const titleContent = [
        genPickerButton(
          this.$createElement,
          displayedHour,
          () => this.$emit('update:selectMode', SelectMode.Hour),
          this.selectMode === SelectMode.Hour,
          this.disabled
        ),
        this.$createElement('span', ':'),
        genPickerButton(
          this.$createElement,
          displayedMinute,
          () => this.$emit('update:selectMode', SelectMode.Minute),
          this.selectMode === SelectMode.Minute,
          this.disabled
        )
      ]

      if (this.useSeconds) {
        const displayedSecond = this.time.second == null ? '--' : pad(this.time.second)
        titleContent.push(this.$createElement('span', ':'))
        titleContent.push(genPickerButton(
          this.$createElement,
          displayedSecond,
          () => this.$emit('update:selectMode', SelectMode.Second),
          this.selectMode === SelectMode.Second,
          this.disabled
        ))
      }
      return this.$createElement('div', {
        'class': 'v-time-picker-title__time'
      }, titleContent)
    },
    genAmPm () {
      return this.$createElement('div', {
        staticClass: 'v-time-picker-title__ampm'
      }, [
        genPickerButton(
          this.$createElement,
          'AM',
          () => this.$emit('update:period', 'am'),
          this.period === 'am',
          this.disabled || this.readonly
        ),
        genPickerButton(
          this.$createElement,
          'PM',
          () => this.$emit('update:period', 'pm'),
          this.period === 'pm',
          this.disabled || this.readonly
        )
      ])
    }
  },

  render (h): VNode {
    const children = [this.genTime()]

    this.isAmPm && this.showAmPm && children.push(this.genAmPm())

    return h('div', {
      staticClass: 'v-time-picker-title'
    }, children)
  }
})
