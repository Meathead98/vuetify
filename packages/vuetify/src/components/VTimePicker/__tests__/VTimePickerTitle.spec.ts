import VTimePickerTitle from '../VTimePickerTitle'
import {
  mount,
  Wrapper,
  MountOptions
} from '@vue/test-utils'
import { SelectMode } from '../VTime'

describe('VTimePickerTitle.ts', () => {
  type Instance = InstanceType<typeof VTimePickerTitle>
  let mountFunction: (options?: MountOptions<Instance>) => Wrapper<Instance>
  beforeEach(() => {
    mountFunction = (options?: MountOptions<Instance>) => {
      return mount(VTimePickerTitle, options)
    }
  })

  it('should render component in 24hr', () => {
    const wrapper = mountFunction({
      propsData: {
        time: {
          hour: 14,
          minute: 13
        },
        format: '24hr'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render component in ampm', () => {
    const wrapper = mountFunction({
      propsData: {
        time: {
          hour: 14,
          minute: 13
        }
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render seconds', () => {
    const wrapper = mountFunction({
      propsData: {
        time: {
          hour: 14,
          minute: 13,
          second: 30
        },
        useSeconds: true
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render ampm buttons', () => {
    const wrapper = mountFunction({
      propsData: {
        time: {
          hour: 14,
          minute: 13
        },
        period: 'am',
        showAmPm: true
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should emit event when clicking hour, minute, second', async () => {
    const selectMode = jest.fn()
    const wrapper = mountFunction({
      propsData: {
        time: {
          hour: 14,
          minute: 13,
          second: 30
        },
        useSeconds: true,
        format: '24hr'
      },
      listeners: {
        'update:selectMode': selectMode
      }
    })

    const buttons = wrapper.findAll('.v-time-picker-title__time .v-picker__title__btn')

    buttons.at(0).element.click()
    await wrapper.vm.$nextTick()

    expect(selectMode).toHaveBeenCalledWith(SelectMode.Hour)

    buttons.at(1).element.click()
    await wrapper.vm.$nextTick()

    expect(selectMode).toHaveBeenCalledWith(SelectMode.Minute)

    buttons.at(2).element.click()
    await wrapper.vm.$nextTick()

    expect(selectMode).toHaveBeenCalledWith(SelectMode.Second)
  })

  it('should render disabled component', () => {
    const wrapper = mountFunction({
      propsData: {
        time: {
          hour: 14,
          minute: 13
        },
        disabled: true
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render with selectMode from mount', () => {
    const wrapper = mountFunction({
      propsData: {
        time: {
          hour: 14,
          minute: 13
        },
        selectMode: SelectMode.Minute
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should emit event when clicking ampm buttons', async () => {
    const period = jest.fn()
    const wrapper = mountFunction({
      propsData: {
        time: {
          hour: 14,
          minute: 13
        },
        period: 'am',
        showAmPm: true
      },
      listeners: {
        'update:period': period
      }
    })

    const buttons = wrapper.findAll('.v-time-picker-title__ampm .v-picker__title__btn')
    buttons.at(1).element.click()
    await wrapper.vm.$nextTick()

    expect(period).toHaveBeenCalledWith('pm')

    wrapper.setProps({
      period: 'pm'
    })
    await wrapper.vm.$nextTick()

    buttons.at(0).element.click()
    await wrapper.vm.$nextTick()

    expect(period).toHaveBeenCalledWith('am')
  })

  it('should not emit event from ampm buttons when readonly', async () => {
    const period = jest.fn()
    const wrapper = mountFunction({
      propsData: {
        time: {
          hour: 14,
          minute: 13
        },
        period: 'am',
        showAmPm: true,
        readonly: true
      },
      listeners: {
        'update:period': period
      }
    })

    const buttons = wrapper.findAll('.v-time-picker-title__ampm .v-picker__title__btn')
    buttons.at(1).element.click()
    await wrapper.vm.$nextTick()

    expect(period).not.toHaveBeenCalled()

    wrapper.setProps({
      period: 'pm'
    })
    await wrapper.vm.$nextTick()

    buttons.at(0).element.click()
    await wrapper.vm.$nextTick()

    expect(period).not.toHaveBeenCalled()
  })

  it('should still emit event when clicking hour, minute, second when readonly', async () => {
    const selectMode = jest.fn()
    const wrapper = mountFunction({
      propsData: {
        time: {
          hour: 14,
          minute: 13,
          second: 30
        },
        readonly: true,
        useSeconds: true,
        format: '24hr'
      },
      listeners: {
        'update:selectMode': selectMode
      }
    })

    const buttons = wrapper.findAll('.v-time-picker-title__time .v-picker__title__btn')

    buttons.at(0).element.click()
    await wrapper.vm.$nextTick()

    expect(selectMode).toHaveBeenCalledWith(SelectMode.Hour)

    buttons.at(1).element.click()
    await wrapper.vm.$nextTick()

    expect(selectMode).toHaveBeenCalledWith(SelectMode.Minute)

    buttons.at(2).element.click()
    await wrapper.vm.$nextTick()

    expect(selectMode).toHaveBeenCalledWith(SelectMode.Second)
  })
})
