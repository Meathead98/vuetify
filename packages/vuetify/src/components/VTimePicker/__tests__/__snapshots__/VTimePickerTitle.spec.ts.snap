// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VTimePickerTitle.ts should render ampm buttons 1`] = `

<div class="v-time-picker-title">
  <div class="v-time-picker-title__time">
    <div class="v-picker__title__btn">
      2
    </div>
    <span>
      :
    </span>
    <div class="v-picker__title__btn">
      13
    </div>
  </div>
  <div class="v-time-picker-title__ampm">
    <div class="v-picker__title__btn v-picker__title__btn--active">
      AM
    </div>
    <div class="v-picker__title__btn">
      PM
    </div>
  </div>
</div>

`;

exports[`VTimePickerTitle.ts should render component in 24hr 1`] = `

<div class="v-time-picker-title">
  <div class="v-time-picker-title__time">
    <div class="v-picker__title__btn">
      14
    </div>
    <span>
      :
    </span>
    <div class="v-picker__title__btn">
      13
    </div>
  </div>
</div>

`;

exports[`VTimePickerTitle.ts should render component in ampm 1`] = `

<div class="v-time-picker-title">
  <div class="v-time-picker-title__time">
    <div class="v-picker__title__btn">
      2
    </div>
    <span>
      :
    </span>
    <div class="v-picker__title__btn">
      13
    </div>
  </div>
</div>

`;

exports[`VTimePickerTitle.ts should render disabled component 1`] = `

<div class="v-time-picker-title">
  <div class="v-time-picker-title__time">
    <div class="v-picker__title__btn v-picker__title__btn--readonly">
      2
    </div>
    <span>
      :
    </span>
    <div class="v-picker__title__btn v-picker__title__btn--readonly">
      13
    </div>
  </div>
</div>

`;

exports[`VTimePickerTitle.ts should render seconds 1`] = `

<div class="v-time-picker-title">
  <div class="v-time-picker-title__time">
    <div class="v-picker__title__btn">
      2
    </div>
    <span>
      :
    </span>
    <div class="v-picker__title__btn">
      13
    </div>
    <span>
      :
    </span>
    <div class="v-picker__title__btn">
      30
    </div>
  </div>
</div>

`;

exports[`VTimePickerTitle.ts should render with selectMode from mount 1`] = `

<div class="v-time-picker-title">
  <div class="v-time-picker-title__time">
    <div class="v-picker__title__btn">
      2
    </div>
    <span>
      :
    </span>
    <div class="v-picker__title__btn v-picker__title__btn--active">
      13
    </div>
  </div>
</div>

`;
